xml-conduit/src/Text/XML/Stream/Parse.hs

```diff
@@ -227,13 +224,31 @@ tokenToEvent _ es n (TokenEndElement name) =
         case n of
             []   -> (NSLevel Nothing Map.empty, [])
             x:xs -> (x, xs)
-tokenToEvent _ es n (TokenContent (ContentEntity e))
-    | Just t <- lookup e es = (es, n, [EventContent $ ContentText t])
+tokenToEvent ps es n (TokenContent (ContentEntity e))
+    = (es, n, map EventContent (resolveEntities 0 ps es [ContentEntity e]))
 tokenToEvent _ es n (TokenContent c) = (es, n, [EventContent c])
 tokenToEvent _ es n (TokenComment c) = (es, n, [EventComment c])
 tokenToEvent _ es n (TokenDoctype t eid es') = (es ++ es', n, [EventBeginDoctype t eid, EventEndDoctype])
 tokenToEvent _ es n (TokenCDATA t) = (es, n, [EventCDATA t])
 
+resolveEntities :: Int             -- recursion depth
+                -> ParseSettings
+                -> [(Text, Text)]  -- entity table
+                -> [Content]
+                -> [Content]
+resolveEntities depth ps es (ContentEntity e : cs)
+    | depth < 10  -- avoid malicious recursive entity expansion
```

Comment:
What about adding a field in `ParseSettings` to control the maximum depth, with default value set to 10 ?

Reply:
I could do that if you want, but it does mean an API change -- do you really want that?

Reply:
Or we could just set this to a higher number, such that no realistic case would exceed it.
(I think 10 already might be high enough, but what about 100?)

Reply:
Another option is to avoid having a limit on recursion depth, but to keep track of the entities we've already expanded in a Set.  If we encounter `bar` in the process of recursively expanding `bar`, then we stop.  That would prevent loops without hard-coding any number.  I'd favor this approach I think.

Another issue that was raised in #103 was the billion laughs attack, e.g.:
```xml
<?xml version="1.0"?>
<!DOCTYPE lolz [
 <!ENTITY lol "lol">
 <!ELEMENT lolz (#PCDATA)>
 <!ENTITY lol1 "&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;">
 <!ENTITY lol2 "&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;">
 <!ENTITY lol3 "&lol2;&lol2;&lol2;&lol2;&lol2;&lol2;&lol2;&lol2;&lol2;&lol2;">
 <!ENTITY lol4 "&lol3;&lol3;&lol3;&lol3;&lol3;&lol3;&lol3;&lol3;&lol3;&lol3;">
 <!ENTITY lol5 "&lol4;&lol4;&lol4;&lol4;&lol4;&lol4;&lol4;&lol4;&lol4;&lol4;">
 <!ENTITY lol6 "&lol5;&lol5;&lol5;&lol5;&lol5;&lol5;&lol5;&lol5;&lol5;&lol5;">
 <!ENTITY lol7 "&lol6;&lol6;&lol6;&lol6;&lol6;&lol6;&lol6;&lol6;&lol6;&lol6;">
 <!ENTITY lol8 "&lol7;&lol7;&lol7;&lol7;&lol7;&lol7;&lol7;&lol7;&lol7;&lol7;">
 <!ENTITY lol9 "&lol8;&lol8;&lol8;&lol8;&lol8;&lol8;&lol8;&lol8;&lol8;&lol8;">
]>
<lolz>&lol9;</lolz>
```
But note that even setting recursion depth to 10 won't prevent this.  I'm not sure preventing this should be a goal of an XML parsing library, since this is just valid XML with a very long expansion.  But if you were worried about this, we could trigger an exception when the expansion of an entity exceeds some fixed length?

Reply:
[This article](https://docs.microsoft.com/en-us/archive/msdn-magazine/2009/november/xml-denial-of-service-attacks-and-defenses) recommends having an option to disable DTD parsing and another option to set the maximum number of characters for entity expansion.  That sounds like a reasonable approach; it would require two additional fields on parser options.

Reply:
> I could do that if you want, but it does mean an API change -- do you really want that?

Constructor of `ParseSettings` isn't exported (see [Text.XML.Stream.Parse#77](https://github.com/snoyberg/xml/blob/master/xml-conduit/src/Text/XML/Stream/Parse.hs#L77)), so it should be backward compatible, unless I am mistaken ?

Even if it weren't, the alternative of having a hardcoded value sounds worse to me, from a user standpoint.

> Another option is to avoid having a limit on recursion depth, but to keep track of the entities we've already expanded in a Set. If we encounter bar in the process of recursively expanding bar, then we stop.

Sounds correct to me, but considering the extra cost in time and memory, in this context, a simpler recursion-depth-based heuristic sounds like a nice tradeoff.

> I'm not sure preventing this should be a goal of an XML parsing library, since this is just valid XML with a very long expansion.

I am not sure it should either, but simple safeguards that don't compromise performance, usability and maintainability are probably fair game. Having a maximum number of characters after expansion seems to qualify.